{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8edcf9e2",
   "metadata": {},
   "source": [
    "---\n",
    "title: Python VCF parser code\n",
    "toc: true\n",
    "description: \n",
    "courses: { csp : {week: 3}}\n",
    "categories: [C3.0, C3.1, C4.1]\n",
    "type: hacks\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5beac902",
   "metadata": {},
   "source": [
    "## VCF python parser I've been using at the lab\n",
    "\n",
    "### Aditional comments are there to understand the code conceptually, it is a bit out of scope for the class but it's a cool project that I've been working on\n",
    "- Unfortunately because the data is supposed to be confidential for my lab, I will not be sharing any of the data sets here\n",
    "- Code utilizes command arguments with the arg parser\n",
    "- Reads and parses genotypes and compares the total amount of differences between them\n",
    "- Additionally, utilizes multiprocessing to read through multiple vcf files, thus why I will not preprocess inputs / outputs like I do with the other lab notebooks because it's just too CPU heavy. However, I will be demoing on a virtual machine.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6842554e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "import concurrent.futures\n",
    "import argparse\n",
    "import time\n",
    "start_time = time.time()\n",
    "\n",
    "def remap_genotype(gt, ref, alt):\n",
    "    # Remap the genotype based on swapped alleles.\n",
    "    if isinstance(gt, tuple):  # If gt is a tuple, convert to string\n",
    "        gt = \"/\".join(str(x) for x in gt)\n",
    "        \n",
    "    gt_fields = gt.split(\"/\")\n",
    "    remapped_fields = []\n",
    "    #Basically takes the number and then it appends the actual letter to the list called remapped_fields\n",
    "    for field in gt_fields:\n",
    "        if field == \"0\":\n",
    "            remapped_fields.append(ref)\n",
    "        elif field == \"1\":\n",
    "            remapped_fields.append(alt)\n",
    "        else:\n",
    "            remapped_fields.append(\".\")\n",
    "\n",
    "    return \"/\".join(remapped_fields)\n",
    "\n",
    "def compare_genotypes(record1, record2, allele_map):\n",
    "    differences = 0\n",
    "\n",
    "    if record1.pos == record2.pos and record1.chrom == record2.chrom:\n",
    "        ref1, alt1 = allele_map.get(record1.id, (record1.ref, record1.alts[0])) # Get record 1's reference and record 1's alternate alleles via their ID\n",
    "        ref2, alt2 = allele_map.get(record2.id, (record2.ref, record2.alts[0])) # Same thing as above but you basically just get it for record 2 nothing too complciated, could possibly made it faster with like use of tabix or something but yeah\n",
    "    \n",
    "\n",
    "        samples1 = list(record1.samples.values())  # Get all samples in file 1\n",
    "        samples2 = list(record2.samples.values())  # Get all samples in file 2\n",
    "\n",
    "        for i in range(len(samples1)):\n",
    "            sample1 = samples1[i]\n",
    "            sample2 = samples2[i]\n",
    "\n",
    "            gt1 = sample1[\"GT\"]\n",
    "            gt2 = sample2[\"GT\"]\n",
    "\n",
    "            # Remap genotypes if references don't match or if the alterante don't match up\n",
    "            if ref1 != ref2 or alt1 != alt2:\n",
    "                gt1 = remap_genotype(gt1, ref1, alt1)\n",
    "                gt2 = remap_genotype(gt2, ref2, alt2)\n",
    "\n",
    "            # Compare the remapped genotypes via the remapped genotypes list which is now represented in G1, G2\n",
    "            if gt1 != gt2:\n",
    "                differences += 1\n",
    "                print(f\"Position: {record1.chrom}:{record1.pos}\")\n",
    "                print(f\"Sample {i}:\")\n",
    "                print(f\"GT1: {gt1}, GT2: {gt2}\")\n",
    "                print(f\"Ref1: {ref1}, Alt1: {alt1}\")\n",
    "                print(f\"Ref2: {ref2}, Alt2: {alt2}\")\n",
    "                print(\"-\" * 40)\n",
    "\n",
    "    return differences\n",
    "\n",
    "def compare_vcf_files(vcf_file1, vcf_file2, dev_mode=False): #If nothing's passed then no dev mode and it'll just pass out the thing\n",
    "    differences = 0\n",
    "\n",
    "    allele_map = {}  # A dictionary to store ref and alt alleles for each record ID\n",
    "    with pysam.VariantFile(vcf_file1) as vcf1, pysam.VariantFile(vcf_file2) as vcf2:\n",
    "        for record in vcf1:\n",
    "            allele_map[record.id] = (record.ref, record.alts[0])  # Assuming single alternate allele\n",
    "\n",
    "        # Use concurrent processing to compare genotypes\n",
    "        with concurrent.futures.ThreadPoolExecutor() as executor:\n",
    "            futures = []\n",
    "            for record1 in vcf1.fetch():\n",
    "                for record2 in vcf2.fetch(record1.chrom, record1.pos, record1.pos + 1):\n",
    "                    future = executor.submit(compare_genotypes, record1, record2, allele_map)\n",
    "                    futures.append(future)\n",
    "\n",
    "            for future in concurrent.futures.as_completed(futures):\n",
    "                differences += future.result()\n",
    "\n",
    "    return differences\n",
    "\n",
    "\n",
    "\n",
    "'''\n",
    "Usage case:\n",
    "\n",
    "python compare.py vcf_file_1 vcf_file_2 --dev_mode\n",
    "\n",
    "Basically dev mode gives u a nice prints so you can see what it thinks/counts as a mutation\n",
    "Right now it does not consider the indel mutations (if for example ref1 does not match up with ref2, but ref2 is longer it will still count it as a mutation for now) but that should be an easy check with the compare_genotypes for length\n",
    "But for now since things have been kinda busy I'm just gonna send what I have right now, since we have to deal with indel mutations later anyway\n",
    "\n",
    "'''\n",
    "if __name__ == \"__main__\":\n",
    "    parser = argparse.ArgumentParser(description=\"Compare two VCF files and find genotype differences.\")\n",
    "    parser.add_argument(\"vcf_file1\", type=str, help=\"Path to the first VCF file\")\n",
    "    parser.add_argument(\"vcf_file2\", type=str, help=\"Path to the second VCF file\")\n",
    "    parser.add_argument(\"--dev_mode\", action=\"store_true\", help=\"Enable developer mode with print statements\")\n",
    "    args = parser.parse_args()\n",
    "\n",
    "    differences = compare_vcf_files(args.vcf_file1, args.vcf_file2, args.dev_mode)\n",
    "\n",
    "    # Time execution (some rando thing I copied from stack overflow to print the time):\n",
    "    end_time = time.time()\n",
    "    total_time = end_time - start_time\n",
    "    minutes = int(total_time // 60)\n",
    "    seconds = total_time % 60\n",
    "    print(\"Number of genotype differences:\", differences)\n",
    "    print(f\"Execution time: {minutes} minutes and {seconds:.2f} seconds\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Before it was commandline parsed\n",
    "# vcf_file1 = \"/home/will/Downloads/mock_data_vcf_1/NDAR-INV119XNUX_S2_merged_L001_markdup_recalibrated_Haplotyper_MOCKDATA.vcf.gz\"\n",
    "# vcf_file2 = \"/home/will/Downloads/mock_data_vcf_1/NDAR-INV1GBYJF8_L001_markdup_recalibrated_Haplotyper_MOCKDATA.vcf.gz\"\n",
    "# differences = compare_vcf_files(vcf_file1, vcf_file2)\n",
    "# print(\"Number of genotype differences:\", differences)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
